src/parser/jobs/sam/Buffs.js

```diff
@@ -0,0 +1,136 @@
+import React, {Fragment} from 'react'
+
+import {ActionLink} from 'components/ui/DbLink'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import Module from 'parser/core/Module'
+import {Rule, Requirement} from '../../core/modules/Checklist'
+import {Suggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+
+const JINPU_DURATION = 30000
+const SHIFU_DURATION = 30000
+
+const BUFF_BUFFER = 10000
+
+export default class Buffs extends Module {
+	static handle = 'buffs'
+	static dependencies = [
+		'checklist',
+		'combatants',
+		'invuln',
+		'suggestions',
+	]
+
+	_JinpuUses = []
+	_ShifuUses = []
+
+	_earlyJinpuApplications = 0
+	_earlyShifyApplications = 0
+
+	constructor(...args) {
+		super(...args)
+
+		const filter = {
+			by: 'player',
+			abilityId: [STATUSES.JINPU.id, STATUSES.SHIFU.id],
+		}
+		this.addHook('applybuff', filter, this._onShifuApplication, this._onJinpuApplication)
+		this.addHook('refreshbuff', filter, this._onShifuApplication, this._onJinpuApplication)
```

Comment:
I'm like 90% sure this doesn't actually call `_onJinpuApplication` ever...

Reply:
I'm pretty sure you need a hook per buff at that point.

Reply:
Yea I think signature of `addHook` is 
```javascript
addHook(events, filter, cb) {
```

Which means the last callback there would be ignored...

Actually maybe a future improvement?
Thoughts @ackwell?

Reply:
Yeah the sig is

```js
addHook(events[, filter], cb)
```

I'm hesitant to allow multiple callbacks per event. The snipping we're commenting on could theoretically be written as

```js
this.addHook(['applybuff', 'refreshbuff'], filter, event => {
	this._onShifuApplication(event)
	this._onJinpuApplication(event)
})
```

You could take it a step further and combine the two callbacks, too - they're doing the same thing.

Reply:
Fair enough. I was just thinking it might be an interesting pattern to consider, splitting the logic in that manner.
But just using a combine function is suitable enough for that...
Maybe provide that boilerplate as a helper function?